<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/Python-Notebook-Banners/Examples.png"  style="display: block; margin-left: auto; margin-right: auto;";/>
</div>

# Examples: Advanced visualisations in Python



In this train, we will use Seaborn and Matplotlib to create a variety of visualisations in Python. We will gain insight into slicing Pandas DataFrames to create advanced visualisations.

## Learning objectives

By the end of this train, you should be able to:
* Navigate a Pandas DataFrame.
* Slice a Pandas DataFrame.
* Produce advanced visualisations in Python.


# Exploring the data

In this notebook, we will learn how to work with Pandas DataFrames and perform some advanced visualisations to draw insights. We'll demonstrate this by working through an example. 

Let's start by importing the libraries we're going to need and loading a dataset about football players and their statistics.

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

In [2]:
soccer = pd.read_csv("https://raw.githubusercontent.com/Explore-AI/Public-Data/master/Data/fundamentals/football_players.csv",low_memory=False)

Let's change the following columns to numerical values to make it possible to perform numerical operations on them.

In [3]:
cols = ['Overall', 'Acceleration', 'Aggression',
       'Agility', 'Balance', 'Ball control', 'Composure', 'Crossing', 'Curve',
       'Dribbling', 'Finishing', 'Free kick accuracy', 'GK diving',
       'GK handling', 'GK kicking', 'GK positioning', 'GK reflexes',
       'Heading accuracy', 'Interceptions', 'Jumping', 'Long passing',
       'Long shots', 'Marking', 'Penalties', 'Positioning', 'Reactions',
       'Short passing', 'Shot power', 'Sliding tackle', 'Sprint speed',
       'Stamina', 'Standing tackle', 'Strength', 'Vision', 'Volleys']

soccer[cols] = soccer[cols].apply(pd.to_numeric, errors='coerce', axis=1)

The following code block shows the first 10 rows of the DataFrame.

In [4]:
soccer.head(10)

Name  Age Nationality  Overall  Acceleration  Aggression  \
0  Cristiano Ronaldo   32    Portugal     94.0          89.0        63.0   
1           L. Messi   30   Argentina     93.0          92.0        48.0   
2             Neymar   25      Brazil     92.0          94.0        56.0   
3          L. Suárez   30     Uruguay     92.0          88.0        78.0   
4           M. Neuer   31     Germany     92.0          58.0        29.0   
5     R. Lewandowski   28      Poland     91.0          79.0        80.0   
6             De Gea   26       Spain     90.0          57.0        38.0   
7          E. Hazard   26     Belgium     90.0          93.0        54.0   
8           T. Kroos   27     Germany     90.0          60.0        60.0   
9         G. Higuaín   29   Argentina     90.0          78.0        50.0   

   Agility  Balance  Ball control  Composure  ...  Short passing  Shot power  \
0     89.0     63.0          93.0       95.0  ...           83.0        94.0   
1     90.0     95.0          95.0       96.0  ...           88.0        85.0   
2     96.0     82.0          95.0       92.0  ...           81.0        80.0   
3     86.0     60.0          91.0       83.0  ...           83.0        87.0   
4     52.0     35.0          48.0       70.0  ...           55.0        25.0   
5     78.0     80.0          89.0       87.0  ...           83.0        88.0   
6     60.0     43.0          42.0       64.0  ...           50.0        31.0   
7     93.0     91.0          92.0       87.0  ...           86.0        79.0   
8     71.0     69.0          89.0       85.0  ...           90.0        87.0   
9     75.0     69.0          85.0       86.0  ...           75.0        88.0   

   Sliding tackle  Sprint speed  Stamina  Standing tackle  Strength  Vision  \
0            23.0          91.0     92.0             31.0      80.0    85.0   
1            26.0          87.0     73.0             28.0      59.0    90.0   
2            33.0          90.0     78.0             24.0      53.0    80.0   
3            38.0          77.0     89.0             45.0      80.0    84.0   
4            11.0          61.0     44.0             10.0      83.0    70.0   
5            19.0          83.0     79.0             42.0      84.0    78.0   
6            13.0          58.0     40.0             21.0      64.0    68.0   
7            22.0          87.0     79.0             27.0      65.0    86.0   
8            69.0          52.0     77.0             82.0      74.0    88.0   
9            18.0          80.0     72.0             22.0      85.0    70.0   

   Volleys  Preferred Positions  
0     88.0               ST LW   
1     85.0                  RW   
2     83.0                  LW   
3     88.0                  ST   
4     11.0                  GK   
5     87.0                  ST   
6     13.0                  GK   
7     79.0                  LW   
8     82.0              CDM CM   
9     88.0                  ST   

[10 rows x 39 columns]

Let's have a look at the descriptive statistics of each column to see if we can spot any anomalies.

In [5]:
soccer.describe()

Age       Overall  Acceleration   Aggression       Agility  \
count  17981.000000  17981.000000  17926.000000  17928.00000  17937.000000   
mean      25.144541     66.247984    103.446893     77.59923     98.150248   
std        4.614272      6.987965    986.386054    768.79730    910.809847   
min       16.000000     46.000000     11.000000     11.00000     14.000000   
25%       21.000000     62.000000     56.000000     43.00000     55.000000   
50%       25.000000     66.000000     67.000000     59.00000     65.000000   
75%       28.000000     71.000000     75.000000     69.00000     74.000000   
max       47.000000     94.000000  33270.000000  43130.00000  30164.000000   

            Balance  Ball control     Composure      Crossing         Curve  \
count  17941.000000  17888.000000  17923.000000  17915.000000  17920.000000   
mean      86.616911    121.420338    103.764995     86.550600     62.586496   
std      752.123941   1236.108966   1028.086556    912.376407    623.794730   
min       11.000000      8.000000      5.000000      5.000000      6.000000   
25%       56.000000     53.000000     51.000000     38.000000     34.000000   
50%       66.000000     63.000000     60.000000     54.000000     48.000000   
75%       74.000000     69.000000     67.000000     64.000000     62.000000   
max    31472.000000  43145.000000  27150.000000  27150.000000  43131.000000   

       ...     Reactions  Short passing    Shot power  Sliding tackle  \
count  ...  17914.000000   17876.000000  17929.000000    17918.000000   
mean   ...    121.637937     115.298557     83.046796       89.598672   
std    ...   1161.367097    1157.803920    811.145271     1074.217147   
min    ...     28.000000      10.000000      3.000000        4.000000   
25%    ...     55.250000      53.000000     46.000000       24.000000   
50%    ...     62.000000      62.000000     59.000000       51.000000   
75%    ...     68.000000      68.000000     69.000000       64.000000   
max    ...  27760.000000   28856.000000  29677.000000    43372.000000   

       Sprint speed       Stamina  Standing tackle      Strength  \
count  17906.000000  17915.000000     17901.000000  17913.000000   
mean     119.034849    123.126151       109.390425    116.040138   
std     1185.940156   1261.551333      1286.912547   1148.520031   
min       11.000000     12.000000         4.000000     12.000000   
25%       57.000000     56.000000        26.000000     58.000000   
50%       67.000000     66.000000        54.000000     66.000000   
75%       75.000000     74.000000        66.000000     74.000000   
max    32905.000000  43433.000000     43144.000000  31868.000000   

             Vision       Volleys  
count  17904.000000  17950.000000  
mean      91.195878     54.622507  
std      945.840135    523.716920  
min       10.000000      4.000000  
25%       43.000000     30.000000  
50%       54.000000     44.000000  
75%       64.000000     57.000000  
max    28216.000000  43129.000000  

[8 rows x 36 columns]

It would appear that we have `17981` entries in total, with some `NaN` values and some values that fall outside of the expected stats range.

We would have difficulty comparing an acceleration score of 67 with an anomaly score of 33270, where we expected all scores to fall below 100.

Let's convert all statistics > 100 to `NaN` and see the summary statistics again.

In [6]:
for column in soccer.describe().columns:
    soccer[column] = soccer[column].apply(lambda x: x if x<=100 else np.nan)

soccer.describe()

Age       Overall  Acceleration    Aggression       Agility  \
count  17981.000000  17981.000000  17897.000000  17913.000000  17910.000000   
mean      25.144541     66.247984     64.576521     55.791492     63.333613   
std        4.614272      6.987965     14.918313     17.476572     14.802800   
min       16.000000     46.000000     11.000000     11.000000     14.000000   
25%       21.000000     62.000000     56.000000     43.000000     55.000000   
50%       25.000000     66.000000     67.000000     59.000000     65.000000   
75%       28.000000     71.000000     75.000000     69.000000     74.000000   
max       47.000000     94.000000     96.000000     96.000000     96.000000   

            Balance  Ball control     Composure      Crossing         Curve  \
count  17924.000000  17840.000000  17887.000000  17885.000000  17908.000000   
mean      63.803894     58.071469     57.877341     49.725301     47.261112   
std       14.100950     16.834554     12.938165     18.474237     18.463394   
min       11.000000      8.000000      5.000000      5.000000      6.000000   
25%       56.000000     53.000000     51.000000     37.000000     34.000000   
50%       66.000000     63.000000     60.000000     54.000000     48.000000   
75%       74.000000     69.000000     67.000000     64.000000     62.000000   
max       96.000000     95.000000     96.000000     91.000000     92.000000   

       ...     Reactions  Short passing    Shot power  Sliding tackle  \
count  ...  17866.000000   17832.000000  17908.000000    17886.000000   
mean   ...     61.898802      58.263964     55.622292       45.496645   
std    ...      9.176709      14.947943     17.408483       21.489926   
min    ...     28.000000      10.000000      3.000000        4.000000   
25%    ...     55.000000      53.000000     46.000000       23.000000   
50%    ...     62.000000      62.000000     59.000000       51.000000   
75%    ...     68.000000      68.000000     69.000000       64.000000   
max    ...     96.000000      92.000000     94.000000       91.000000   

       Sprint speed       Stamina  Standing tackle      Strength  \
count  17867.000000  17873.000000     17857.000000  17877.000000   
mean      64.795713     63.233201        47.352187     65.279130   
std       14.649127     15.942738        21.851023     12.618415   
min       11.000000     12.000000         4.000000     12.000000   
25%       57.000000     56.000000        26.000000     58.000000   
50%       67.000000     66.000000        54.000000     66.000000   
75%       75.000000     74.000000        66.000000     74.000000   
max       96.000000     95.000000        92.000000     98.000000   

             Vision       Volleys  
count  17874.000000  17940.000000  
mean      52.967495     43.198495  
std       14.387015     17.734259  
min       10.000000      4.000000  
25%       43.000000     30.000000  
50%       54.000000     44.000000  
75%       64.000000     57.000000  
max       94.000000     91.000000  

[8 rows x 36 columns]

Now we have a much better dataset with all the anomalies removed, so we can easily compare the values that are available to us.

## Slicing

Slicing is important for extracting a subset of a DataFrame. Using ``.iloc[]``, we will be able to slice the DataFrame similarly to the way we would slice a list. It uses number slicing to get **positional** records. The syntax for slicing is as follows:

```
dataframe.iloc[start_row:stop_row , start_column:stop_column]
``` 

When slicing in Pandas the start bound is included in the output, i.e. if we slice rows `[0:5]`, we will extract rows 0 to 4. 

Another slicing method is to use the ```.loc[]``` function native to Pandas. `.loc[]` allows us to slice based on data _labels_, as opposed to `.iloc[]` which slices based on the DataFrame _index_. An easy way to remember the difference is the **i** in `.iloc[]` is for **i**ndex.

In [7]:
soccer.iloc[1:101,:] # selects rows 1 to 100 from the DataFrame

Name  Age Nationality  Overall  Acceleration  Aggression  \
1          L. Messi   30   Argentina     93.0          92.0        48.0   
2            Neymar   25      Brazil     92.0          94.0        56.0   
3         L. Suárez   30     Uruguay     92.0          88.0        78.0   
4          M. Neuer   31     Germany     92.0          58.0        29.0   
5    R. Lewandowski   28      Poland     91.0          79.0        80.0   
..              ...  ...         ...      ...           ...         ...   
96          D. Alli   21     England     84.0          77.0        84.0   
97          S. Mané   25     Senegal     84.0          94.0        73.0   
98         Carvajal   25       Spain     84.0          76.0        81.0   
99       J. Draxler   23     Germany     84.0          76.0        48.0   
100       M. Icardi   24   Argentina     84.0          75.0        55.0   

     Agility  Balance  Ball control  Composure  ...  Short passing  \
1       90.0     95.0          95.0       96.0  ...           88.0   
2       96.0     82.0          95.0       92.0  ...           81.0   
3       86.0     60.0          91.0       83.0  ...           83.0   
4       52.0     35.0          48.0       70.0  ...           55.0   
5       78.0     80.0          89.0       87.0  ...           83.0   
..       ...      ...           ...        ...  ...            ...   
96      74.0     62.0          85.0       86.0  ...           83.0   
97      89.0     85.0          84.0       78.0  ...           77.0   
98      82.0     79.0          83.0       74.0  ...           81.0   
99      81.0     75.0          87.0       81.0  ...           84.0   
100     74.0     76.0          77.0       82.0  ...           73.0   

     Shot power  Sliding tackle  Sprint speed  Stamina  Standing tackle  \
1          85.0            26.0          87.0     73.0             28.0   
2          80.0            33.0          90.0     78.0             24.0   
3          87.0            38.0          77.0     89.0             45.0   
4          25.0            11.0          61.0     44.0             10.0   
5          88.0            19.0          83.0     79.0             42.0   
..          ...             ...           ...      ...              ...   
96         77.0            57.0          76.0     89.0             63.0   
97         77.0            38.0          92.0     82.0             35.0   
98         64.0            86.0          86.0     84.0             85.0   
99         86.0            26.0          79.0     68.0             35.0   
100        81.0            21.0          80.0     70.0             21.0   

     Strength  Vision  Volleys  Preferred Positions  
1        59.0    90.0     85.0                  RW   
2        53.0    80.0     83.0                  LW   
3        80.0    84.0     88.0                  ST   
4        83.0    70.0     11.0                  GK   
5        84.0    78.0     87.0                  ST   
..        ...     ...      ...                  ...  
96       71.0    83.0     76.0           LM CF CAM   
97       72.0    78.0     68.0               LW RW   
98       75.0    57.0     34.0                  RB   
99       67.0    84.0     84.0           LM CAM LW   
100      74.0    64.0     81.0                  ST   

[100 rows x 39 columns]

## Renaming columns

In Python, it is good practice to label columns in a way that conforms to PEP 8 naming standards (for example, no spaces in the name). 

Below we illustrate how to change the name of a column in a Pandas DataFrame.

In [8]:
soccer = soccer.rename(columns = {"Preferred Positions": "Preferred_Positions"})
soccer.head()

Name  Age Nationality  Overall  Acceleration  Aggression  \
0  Cristiano Ronaldo   32    Portugal     94.0          89.0        63.0   
1           L. Messi   30   Argentina     93.0          92.0        48.0   
2             Neymar   25      Brazil     92.0          94.0        56.0   
3          L. Suárez   30     Uruguay     92.0          88.0        78.0   
4           M. Neuer   31     Germany     92.0          58.0        29.0   

   Agility  Balance  Ball control  Composure  ...  Short passing  Shot power  \
0     89.0     63.0          93.0       95.0  ...           83.0        94.0   
1     90.0     95.0          95.0       96.0  ...           88.0        85.0   
2     96.0     82.0          95.0       92.0  ...           81.0        80.0   
3     86.0     60.0          91.0       83.0  ...           83.0        87.0   
4     52.0     35.0          48.0       70.0  ...           55.0        25.0   

   Sliding tackle  Sprint speed  Stamina  Standing tackle  Strength  Vision  \
0            23.0          91.0     92.0             31.0      80.0    85.0   
1            26.0          87.0     73.0             28.0      59.0    90.0   
2            33.0          90.0     78.0             24.0      53.0    80.0   
3            38.0          77.0     89.0             45.0      80.0    84.0   
4            11.0          61.0     44.0             10.0      83.0    70.0   

   Volleys  Preferred_Positions  
0     88.0               ST LW   
1     85.0                  RW   
2     83.0                  LW   
3     88.0                  ST   
4     11.0                  GK   

[5 rows x 39 columns]

## Working with data types

Thinking back to previous lessons, we will remember that Python has different data types, each with their own characteristics and rules. We may remember that categorical data can have _grouping_ or _order_ and do not need to be (but can be) numerical. Examples would be:
* Sex 
* Country of origin


To find non-numerical categorical data in our DataFrame, we first use ```.select_dtypes(include/exclude)``` to find the data type. 

In Python, non-numerical data are usually stored as an ```object```. Let's identify the non-numeric data columns in the ```soccer``` DataFrame.

In [ ]:
soccer.select_dtypes(include=['object']).head()

In some cases, it might benefit our analysis to change the data types in our DataFrame (such as converting an `integer` to a `float`). The code below enables us to do this. 

Let's take ```Age```, generally expressed as an `int`, and convert it to a `float` using the ```.astype()``` function.

In [ ]:
soccer['Age'] = soccer['Age'].astype('float')
soccer['Age'].dtype

# Visualising the data

Pandas DataFrames are a great way to wrangle data, but to see trends and patterns, find key insights, and identify outliers in the data, it generally helps to visualise them in a meaningful way. Python has some of the best graphing packages available to facilitate this. ```Matplotlib``` and ```Seaborn``` are two of the most commonly used packages for a variety of visualisations. 

Let's demonstrate some useful visualisations.

In [ ]:
# set our colour palette preference for seaborn
sns.set_palette("Spectral_r",n_colors=4)

## Histogram

A histogram is very similar to a bar chart, except that it **groups** data or puts it into **ranges**. These groupings/ranges are called "**bins**".

```Seaborn``` and ```Matplotlib``` have their own histogram functions. 

```Seaborn``` will combine the histogram with a distribution if prompted to do so.
If we want to display the distribution, we simply add the argument ```kde=True``` to the plot.

Let's look at how to graph a histogram in both `Matplotlib` and `Seaborn` below.

In [ ]:
#seaborn
fig, ax = plt.subplots(2,1, figsize=(10,8))

sns.histplot(soccer['Age'], bins=5, element="step", ax=ax[0])
sns.histplot(soccer['Age'], bins=5, element="step", kde=True, ax=ax[1])

ax[0].set_title('Histogram for Ages')
ax[1].set_title('Histogram of Ages with distribution')

plt.tight_layout()
plt.show()

In [ ]:
#matplotlib
plt.xlabel('Age')
plt.hist(soccer['Age'], bins=5)
plt.show()

Note how ```Seaborn``` automatically labels the axes for us whereas ```Matplotlib``` does not and the axes need to be specified.

## Jointplots

```Seaborn```'s _jointplot_ will graphically display two variables on one plot and a single variable on the marginal sides of the graph. Essentially, a _jointplot_ allows us to see the relationship between two variables as well as each variable independently. 

Let's see an example below: 

In [ ]:
j = sns.jointplot(data=soccer, x='Overall', y='Age', kind='reg')
plt.suptitle('Jointplot showing the Relationship between Overall and Age of Football Players', y=1.1)
plt.show()

In the graph above, it is still difficult to see how the `Overall` and `Age` variables are related. To improve this, we can try changing the type of plot or _kind_ to a hexagon, which displays the number of matching points much more clearly. 

Looking at the code below, we added the argument ```kind='hex'```.

In [ ]:
j = sns.jointplot(data=soccer, x='Overall', y='Age', kind='hex')
plt.suptitle('Jointplot showing the Relationship between Overall and Age of Football Players', y=1.1)
plt.show()

## Box plots

Box plots are used to show the spread of data. The main difference between a box plot and a histogram (which also shows a distribution as mentioned above) is that it takes up less space to plot. 

Box plots are constructed using the 5-number summary as indicated below:

```{Lower Extreme ; Q1 ; Median ; Q3 ; Upper Extreme}```

![](https://upload.wikimedia.org/wikipedia/commons/5/5c/Labeled_Box_and_Whisker_Plot.png)

Let's see how the `Overall` stats compare based on different ages (from 23 to 26 years old):


In [ ]:
young_players = soccer[(soccer['Age'] > 22) & (soccer['Age'] < 27)] #filter ages

In [ ]:
plt.figure(figsize=(18,5))

sns.boxplot(data=young_players, x='Age', y='Overall')
plt.show()

From the plot above, we can see the general trends of `Overall` stats for each specific age. Box plots, compared to histograms, can make it much easier to compare groupings of data.

## Violin plot

Violin plots are similar to box plots, but they use the KDE or distribution of values for groupings, in addition to a box (the box is still present within each graph). 

Let's plot one using ```Seaborn``` with the `Overall` stats for various `Age` categories (from 23 to 26 years old).

In [ ]:
plt.figure(figsize=(18,5))

sns.violinplot(data=young_players, x='Age', y='Overall')
plt.show()

The violin plots show us the general distribution across each age, and it makes it much easier for us to compare them.

## Facet grid

Sometimes a useful approach (specifically when exploring medium-dimensional data) is to draw the same plot (a facet) on multiple subsets of the dataset, resulting in a matrix of visualisations (a facet grid). This technique is commonly referred to as “lattice”, or “trellis” plotting, and it is related to the idea of _small multiples_. 

Our data will need to be in a Pandas DataFrame for this to work.

Let's try this on our dataset.

In [ ]:
goalkeepers = soccer[(soccer['Preferred_Positions'].str.contains('GK'))]

In [ ]:
sns.pairplot(goalkeepers[['Age','GK diving','GK handling', 'GK kicking', 'GK positioning', 'GK reflexes']],
             hue='Age',palette="Spectral_r")
plt.show()

To practice further, we will work with one of the practice datasets in ```Seaborn```.

In [ ]:
tips = sns.load_dataset("tips")

In [ ]:
g = sns.FacetGrid(tips, col="day")
g.map(plt.hist, "total_bill");
plt.show()

Let's plot another example.

In [ ]:
g = sns.FacetGrid(tips, col="day", row="sex")
g.map(plt.hist, "total_bill");
plt.show()

# Summary

Data visualisation is the act of taking information (data) and placing it into a visual context – such as a map or graph – to make sense of it. Data visualisations make it simpler for the human brain to understand data, making it easier to detect patterns, trends, and outliers.

In this train, we learned how to navigate a Pandas DataFrame, how to slice a Pandas DataFrame, and how to produce advanced visualisations in Python.

After completing this train, we should have a very good idea of how to effectively visualise  data with ```Seaborn``` and ```Matplotlib```. However, this train is by no means exhaustive, so it is encouraged to experiment and explore the many other plots available.


#  

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/refs/heads/master/ALX_banners/ALX_Navy.png"  style="width:140px";/>
</div>